File: frontend/src/utils/encryption.js
import CryptoJS from "crypto-js";

const ENCRYPTION_KEY = import.meta.env.VITE_ENCRYPTION_KEY || "0123456789abcdef0123456789abcdef";

/**
 * Encrypt plaintext using AES. Returns base64 string (IV prepended).
 */
export function encryptData(plaintext) {
  const key = CryptoJS.enc.Hex.parse(ENCRYPTION_KEY);
  const iv = CryptoJS.lib.WordArray.random(16);
  const encrypted = CryptoJS.AES.encrypt(plaintext, key, {
    iv: iv,
    mode: CryptoJS.mode.CBC,
    padding: CryptoJS.pad.Pkcs7,
  });
  // Prepend IV to ciphertext and base64 encode
  const combined = iv.concat(encrypted.ciphertext);
  return CryptoJS.enc.Base64.stringify(combined);
}

/**
 * Decrypt a base64 AES-CBC ciphertext (IV prepended).
 */
export function decryptData(ciphertextB64) {
  const key = CryptoJS.enc.Hex.parse(ENCRYPTION_KEY);
  const raw = CryptoJS.enc.Base64.parse(ciphertextB64);
  const iv = CryptoJS.lib.WordArray.create(raw.words.slice(0, 4), 16);
  const ciphertext = CryptoJS.lib.WordArray.create(raw.words.slice(4), raw.sigBytes - 16);
  const decrypted = CryptoJS.AES.decrypt({ ciphertext: ciphertext }, key, {
    iv: iv,
    mode: CryptoJS.mode.CBC,
    padding: CryptoJS.pad.Pkcs7,
  });
  return decrypted.toString(CryptoJS.enc.Utf8);
}

File: backend/utils/encryption.py
import base64
import os
from Crypto.Cipher import AES
from Crypto.Util.Padding import pad, unpad
from config import ENCRYPTION_KEY


def get_key() -> bytes:
    return bytes.fromhex(ENCRYPTION_KEY)


def encrypt(plaintext: str) -> str:
    """Encrypt a string using AES-CBC and return base64-encoded ciphertext."""
    key = get_key()
    iv = os.urandom(16)
    cipher = AES.new(key, AES.MODE_CBC, iv)
    padded = pad(plaintext.encode("utf-8"), AES.block_size)
    encrypted = cipher.encrypt(padded)
    # Prepend IV to ciphertext, then base64 encode
    return base64.b64encode(iv + encrypted).decode("utf-8")


def decrypt(ciphertext_b64: str) -> str:
    """Decrypt a base64-encoded AES-CBC ciphertext."""
    key = get_key()
    raw = base64.b64decode(ciphertext_b64)
    iv = raw[:16]
    encrypted = raw[16:]
    cipher = AES.new(key, AES.MODE_CBC, iv)
    decrypted = unpad(cipher.decrypt(encrypted), AES.block_size)
    return decrypted.decode("utf-8")
